File: app/src/main/java/top/witzzz/moonbazaar/data/MoonRepository.kt
package top.witzzz.moonbazaar.data

import org.json.JSONObject

// Lightweight typed/call wrappers around the Worker API.
// Every call talks to the Worker so the app never holds the MoonBazaar API key
// or the user token; Session is sent as X-Client-Session by MoonWorkerApi.
object MoonRepository {

    private val api = MoonWorkerApi()

    data class SessionStatus(
        val okResponse: Boolean,
        val sessionValid: Boolean,
        val hasTokens: Boolean,
        val uid: String?,
        val apiKeyConfigured: Boolean,
        val rawText: String
    )

    // /status
    fun status(): Pair<SessionStatus?, String> {
        val r = api.get("/status")
        if (!r.ok) return null to (r.readable.ifEmpty { "HTTP ${r.httpStatus}" })
        val j = r.json ?: return null to "无法解析返回 JSON"
        return SessionStatus(
            okResponse = j.optBoolean("ok", false),
            sessionValid = j.optBoolean("session_valid", false),
            hasTokens = j.optBoolean("has_tokens", false),
            uid = j.optString("uid").ifBlank { null },
            apiKeyConfigured = j.optBoolean("api_key_configured", false),
            rawText = r.body
        ) to ""
    }

    // /proxy/account/*
    fun accountStatus(): Pair<JSONObject?, String> = proxyObj("account/status")
    fun accountBalance(): Pair<JSONObject?, String> = proxyObj("account/balance")

    // /proxy/transactions
    fun transactions(limit: Int = 30, offset: Int = 0): Pair<JSONObject?, String> =
        proxyObj("transactions?limit=$limit&offset=$offset")

    // /proxy/referrals and /proxy/referrals/link
    fun referrals(limit: Int = 50, offset: Int = 0): Pair<JSONObject?, String> =
        proxyObj("referrals?limit=$limit&offset=$offset")
    fun referralLink(): Pair<JSONObject?, String> = proxyObj("referrals/link")

    // /proxy/products (withdraw products)
    fun products(): Pair<JSONObject?, String> = proxyObj("products")

    // /proxy/surveys (main surveys). Worker auto-fills client_ip.
    fun surveys(): Pair<JSONObject?, String> = proxyObj("surveys")

    data class WriteResult(val body: String, val http: Int, val okHttp: Boolean) {
        val json: JSONObject?
            get() = try {
                if (body.isBlank()) null else JSONObject(body)
            } catch (_: Exception) {
                null
            }
        val errorMsg: String
            get() = json?.optJSONObject("error")?.optString("message")
                ?.takeIf { it.isNotBlank() }
                ?: (if (okHttp) "" else "HTTP $http")
    }

    // POST /proxy/withdrawals (idempotency-required)
    fun submitWithdrawal(
        productId: Long,
        amount: String?,
        deliveryDetails: String,
        note: String?
    ): WriteResult {
        val b = JSONObject().apply {
            put("product_id", productId)
            if (!amount.isNullOrBlank()) put("amount", amount.trim())
            put("delivery_details", deliveryDetails)
            if (!note.isNullOrBlank()) put("note", note)
        }
        return write("withdrawals", b)
    }

    // POST /proxy/codes/redeem (idempotency-required); usually HTTP 202,
    // returns a confirmation_url the user must open to finalise.
    fun redeemCode(code: String): WriteResult {
        val b = JSONObject().put("code", code.trim())
        return write("codes/redeem", b)
    }

    // POST /proxy/account/email-verification (no idempotency header).
    fun requestEmailVerification(): WriteResult {
        val b = JSONObject()
        return write("account/email-verification", b, needIdem = false)
    }

    // DELETE /user/tokens: revoke WORKER session server-side.
    fun logout(): WriteResult {
        val r = api.del("/user/tokens")
        return WriteResult(body = r.body, http = r.httpStatus, okHttp = r.ok)
    }

    private fun proxyObj(path: String): Pair<JSONObject?, String> {
        val r = api.proxyGet(path)
        if (!r.ok) return null to (r.readable.ifEmpty { "HTTP ${r.httpStatus}" })
        val j = r.json ?: return null to "上游返回空或非 JSON"
        return j to ""
    }

    private fun write(path: String, body: JSONObject, needIdem: Boolean = true): WriteResult {
        val res = if (needIdem) api.proxyPost(path, body, genIdem(path))
        else api.proxyPostNoIdem(path, body)
        return WriteResult(body = res.body, http = res.httpStatus, okHttp = res.ok)
    }

    private var seq = 0L
    fun genIdem(prefix: String = "op"): String {
        val ts = System.currentTimeMillis()
        seq = (seq + 1) % 100000
        return "$prefix-$ts-$seq-${Session.tokenValue?.take(8) ?: "anon"}"
    }

    // Render a GC (Gascoin) value as an integer string: drop the decimals.
    fun gcInteger(value: Any?): String {
        if (value == null) return "0"
        val d: Double = when (value) {
            is Number -> value.toDouble()
            is String -> value.trim().toDoubleOrNull()
            else -> null
        } ?: return "0"
        if (d < 0) return "0"
        return kotlin.math.floor(d).toLong().toString()
    }
}

File: ios/MoonBazaar/MoonRepository.swift
import Foundation

/// Worker 各接口的封装与幂等键生成（对应 Android 的 `data/MoonRepository.kt`）。
///
/// 每次调用都走 Worker，因此 App 既不持有 MoonBazaar API Key，也不持有用户 token；
/// `Session` 由 `MoonWorkerApi` 以 `X-Client-Session` 头发送。
@MainActor
enum MoonRepository {

    private static let api = MoonWorkerApi()

    // MARK: - /status

    struct SessionStatus {
        var okResponse: Bool = false
        var sessionValid: Bool = false
        var hasTokens: Bool = false
        var uid: String?
        var apiKeyConfigured: Bool = false
        var rawText: String = ""
    }

    /// 返回 (状态, 错误信息)。状态为 nil 时看错误信息。
    static func status() async -> (SessionStatus?, String) {
        let r = await api.get("/status")
        guard r.ok else { return (nil, r.readable.isEmpty ? "HTTP \(r.httpStatus)" : r.readable) }
        guard let j = r.json else { return (nil, "无法解析返回 JSON") }
        let s = SessionStatus(
            okResponse: J.bool(j, "ok"),
            sessionValid: J.bool(j, "session_valid"),
            hasTokens: J.bool(j, "has_tokens"),
            uid: J.str(j, "uid").isEmpty ? nil : J.str(j, "uid"),
            apiKeyConfigured: J.bool(j, "api_key_configured"),
            rawText: r.body
        )
        return (s, "")
    }

    // MARK: - /proxy/account/*

    static func accountStatus() async -> (JObj?, String) { await proxyObj("account/status") }
    static func accountBalance() async -> (JObj?, String) { await proxyObj("account/balance") }

    // MARK: - 只读列表

    static func transactions(limit: Int = 30, offset: Int = 0) async -> (JObj?, String) {
        await proxyObj("transactions?limit=\(limit)&offset=\(offset)")
    }

    static func referrals(limit: Int = 50, offset: Int = 0) async -> (JObj?, String) {
        await proxyObj("referrals?limit=\(limit)&offset=\(offset)")
    }

    static func referralLink() async -> (JObj?, String) { await proxyObj("referrals/link") }

    /// 提现商品列表。
    static func products() async -> (JObj?, String) { await proxyObj("products") }

    /// 主要问卷。Worker 会自动补 client_ip。
    static func surveys() async -> (JObj?, String) { await proxyObj("surveys") }

    // MARK: - 写操作

    struct WriteResult {
        var body: String
        var http: Int
        var okHttp: Bool

        var json: JObj? {
            guard !body.isEmpty, let data = body.data(using: .utf8) else { return nil }
            return (try? JSONSerialization.jsonObject(with: data)) as? JObj
        }

        var errorMsg: String {
            let fallback = okHttp ? "" : "HTTP \(http)"
            guard let j = json, let rawError = j["error"] else { return fallback }
            if let m = J.dict(rawError)?["message"] as? String, !m.isEmpty { return m }
            return fallback
        }
    }

    /// POST /proxy/withdrawals（需要幂等键）。
    static func submitWithdrawal(productId: Int64,
                                 amount: String?,
                                 deliveryDetails: String,
                                 note: String?) async -> WriteResult {
        var b: JObj = [:]
        b["product_id"] = productId
        if let amount = amount, !amount.trimmingCharacters(in: .whitespaces).isEmpty {
            b["amount"] = amount.trimmingCharacters(in: .whitespaces)
        }
        b["delivery_details"] = deliveryDetails
        if let note = note, !note.trimmingCharacters(in: .whitespaces).isEmpty { b["note"] = note }
        return await write("withdrawals", b)
    }

    /// POST /proxy/codes/redeem（需要幂等键）；通常返回 202 与 `confirmation_url`。
    static func redeemCode(_ code: String) async -> WriteResult {
        let b: JObj = ["code": code.trimmingCharacters(in: .whitespacesAndNewlines)]
        return await write("codes/redeem", b)
    }

    /// POST /proxy/account/email-verification（不带幂等键）。
    static func requestEmailVerification() async -> WriteResult {
        await write("account/email-verification", [:], needIdem: false)
    }

    /// DELETE /user/tokens：吊销 Worker session。
    static func logout() async -> WriteResult {
        let r = await api.del("/user/tokens")
        return WriteResult(body: r.body, http: r.httpStatus, okHttp: r.ok)
    }

    // MARK: - 内部

    private static func proxyObj(_ path: String) async -> (JObj?, String) {
        let r = await api.proxyGet(path)
        guard r.ok else { return (nil, r.readable.isEmpty ? "HTTP \(r.httpStatus)" : r.readable) }
        guard let j = r.json else { return (nil, "上游返回空或非 JSON") }
        return (j, "")
    }

    private static func write(_ path: String, _ body: JObj, needIdem: Bool = true) async -> WriteResult {
        let res: MoonWorkerApi.CallResult
        if needIdem {
            res = await api.proxyPost(path, body, idempotencyKey: genIdem(path))
        } else {
            res = await api.proxyPostNoIdem(path, body)
        }
        return WriteResult(body: res.body, http: res.httpStatus, okHttp: res.ok)
    }

    private static var seq: Int64 = 0
    static func genIdem(_ prefix: String = "op") -> String {
        let ts = Int64(Date().timeIntervalSince1970 * 1000)
        seq = (seq + 1) % 100000
        let sessPrefix = Session.shared.tokenValue.map { String($0.prefix(8)) } ?? "anon"
        return "\(prefix)-\(ts)-\(seq)-\(sessPrefix)"
    }

    /// 把 GC（Gascoin）渲染为整数串：丢掉小数。
    static func gcInteger(_ value: Any?) -> String {
        guard let d = J.anyDouble(value) else { return "0" }
        if d < 0 || d.isNaN { return "0" }
        return String(Int64(d.rounded(.down)))
    }
}